api/server.go

```diff
@@ -560,19 +569,51 @@ func startServer(handler http.Handler) {
 		if err != nil {
 			fatal(err)
 		}
-		fmt.Printf("tsuru HTTP/TLS server listening at %s...\n", listen)
-		err = srv.ListenAndServeTLS(certFile, keyFile)
-	} else {
-		fmt.Printf("tsuru HTTP server listening at %s...\n", listen)
-		err = srv.ListenAndServe()
+	} else if listen == "" {
+		fatal(errors.New(`missing "listen" config key`))
 	}
-	if err != nil {
-		fmt.Printf("Listening stopped: %s\n", err)
-		if errOp, ok := err.(*net.OpError); ok {
-			if errOp.Op == "listen" {
-				os.Exit(1)
-			}
+
+	readTimeout, _ := config.GetInt("server:read-timeout")
+	writeTimeout, _ := config.GetInt("server:write-timeout")
+	if listen != "" {
+		httpSrv = &http.Server{
+			ReadTimeout:  time.Duration(readTimeout) * time.Second,
+			WriteTimeout: time.Duration(writeTimeout) * time.Second,
+			Addr:         listen,
+			Handler:      handler,
+		}
+		shutdown.Register(httpSrv)
+	}
+	if tlsListen != "" {
+		httpsSrv = &http.Server{
+			ReadTimeout:  time.Duration(readTimeout) * time.Second,
+			WriteTimeout: time.Duration(writeTimeout) * time.Second,
+			Addr:         tlsListen,
+			Handler:      handler,
 		}
+		shutdown.Register(httpsSrv)
 	}
-	<-shutdownChan
+
+	if httpsSrv != nil {
+		if httpSrv != nil {
+			go startHttpServer(httpSrv, listen)
```

Comment:
By starting a goroutine here we are ignoring possible errors during listening. Maybe we should have a channel for errors to capture then, something like:

```
errCh := make(chan error, 2)
if httpsSrv != nil {
  go func() { errCh <- startHttpsServer(httpsSrv, listen, ...) }()
}
if httpSrv != nil {
  go func() { errCh <- startHttpServer(httpSrv, listen) }()
}
return <-errCh
```

Also, I don't think we should call `fatal` in `createServers` since we're able to return errors. We should simply return the errors and let the caller call `fatal` for any returned error.